tests/Insightly/Listeners/ActivateIntegrationWithCouponTest.php

```diff
@@ -0,0 +1,172 @@
+    {
+        // Given
+        $insightlyOpportunityId = 42;
+        $insightlyProjectId = 51;
+        $integration = $this->givenThereIsAnIntegrationWithId($this->integrationId);
+        $this->givenThereIsAnOpportunityMapping($this->integrationId);
+        $coupon = $this->givenThereIsACoupon($this->integrationId);
+
+        $this->integrationRepository->activateWithCouponCode($this->integrationId, $this->couponCode);
+
+        // Then
+        // It updates the stage of the opportunity
+        $this->opportunityResource->expects($this->once())
+            ->method('updateStage')
+            ->with($insightlyOpportunityId, OpportunityStage::CLOSED);
+
+        // It updates the state of the opportunity
+        $this->opportunityResource->expects($this->once())
+            ->method('updateState')
+            ->with($insightlyOpportunityId, OpportunityState::WON);
+
+        // It creates the project at Insightly
+        $this->projectResource->expects($this->once())
+            ->method('create')
+            ->with($integration)
+            ->willReturn($insightlyProjectId);
+
+        // It updates the project with a coupon code
+        $this->projectResource->expects($this->once())
+            ->method('updateWithCoupon')
+            ->with($insightlyProjectId, $coupon->code);
+
+        // It stores the insightlyProjectId mapping
+        $insightlyMapping = new InsightlyMapping(
+            $this->integrationId,
+            $insightlyProjectId,
+            ResourceType::Project,
+        );
+
+        $this->insightlyMappingRepository->expects(self::once())
+            ->method('save')
+            ->with($insightlyMapping);
+
+        // TODO: Test the correct state and stage on project
+
+        // It links the opportunity to the project
+        $this->projectResource->expects($this->once())
+            ->method('linkOpportunity')
+            ->with($insightlyProjectId, $insightlyOpportunityId);
+
+        // TODO: Test the contacts
+
+        // When
+        $event = new IntegrationActivatedWithCoupon($this->integrationId);
+        $this->listener->handle($event);
+    }
+
+    private function givenThereIsAnIntegrationWithId(UuidInterface $integrationId): Integration
+    {
+        $integration = new Integration(
+            $this->integrationId,
+            IntegrationType::SearchApi,
+            'My integration',
+            'This is my integration',
+            Uuid::uuid4(),
+            IntegrationStatus::Draft,
+            []
+        );
+
+        $this->integrationRepository->expects($this->exactly(1))
+            ->method('getById')
+            ->with($integrationId)
+            ->willReturn($integration);
+
+        return $integration;
+    }
+
+    private function givenThereIsAnOpportunityMapping(UuidInterface $integrationId): InsightlyMapping
+    {
+        $insightlyMapping = new InsightlyMapping(
+            $integrationId,
+            42,
```

Comment:
Maybe not bad to make this Insightly id a parameter. Later we should also extract these given methods to a reusable component like a class or trait.

Reply:
Solved in https://github.com/cultuurnet/publiq-platform/pull/111/commits/ea604ce6c45cfc2c07f438be60298cbd1fe6011d